### DISCUSSION TITLE
Can we build a static library of an App and easily used it in other computers with the same OS

### DISCUSSION THREAD

**Discussion by zx1987** [idaholab/moose#22653] [Q&A General] [answered]
Dear Moose user group,

I was hoping to compile a static library of an App, to use it on different computers with the same OS by just copying that compiled app. I was searching and found this discussion: https://github.com/idaholab/moose/discussions/22326

I was thinking maybe I can build libmesh and moose statically as described in the above post, and then compile our app after that. Will this give us a static library of the app, or there are additional steps I should consider? Thank you!

Best,
Xiang

**Comment by GiudGiud** [comment, marked as answer]
You'll have to follow the post you linked to, then add the static flags in the CXXFLAGS environment variable.
@milljm

**Reply by milljm** [reply]
I agree, the post you linked, is basically the gist of what needs to happen. MOOSE based apps only build according to how libMesh was built. But if you're after a total relocatable binary, you may need to build statically, starting with PETSc.

**Reply by zx1987** [reply]
Thank you @milljm @GiudGiud for your confirmation. I will test this out and let you know. Thanks.

**Reply by zx1987** [reply]
I tried to do the static build of petsc, but came across the error like "/home/xzhang16/mambaforge3/envs/petsc/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: cannot find -lhdf5_cpp: No such file or directory"

Do you have any idea what might be wrong? @milljm @GiudGiud  I have more detailed error message attached in case it could be helpful. Thank you!
[StaticLibmeshError.txt](https://github.com/idaholab/moose/files/9991070/StaticLibmeshError.txt)

**Reply by GiudGiud** [reply]
which HDF5 library are you trying to use? One provided by your local environment or by PETSc ?

**Reply by zx1987** [reply]
I did not really specify which hdf5 to use and was assuming if I take the stand installation procedure, it should be using the one from PETSc.  So I just did this again, by resetting my $PATH so the system only have access to the mamabaforge3 folder hence libraries from PETSc. It turns out I had the same error.  I tried to search lhdf5 in the mambaforge3 and nothing was found.  Tere are the steps I took in this new attempt: 
```
`curl -L -O https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-Linux-x86_64.sh`
`bash Mambaforge-Linux-x86_64.sh -b -p ~/mambaforge3`
`unset LD_LIBRARY_PATH`
`PATH=$(getconf PATH)` 
`export PATH=$HOME/mambaforge3/bin:$PATH  # reset the PATH such that I will not mistakenly use things other than from the the mambaforge3 folder`
`conda config --add channels https://conda.software.inl.gov/public`

steps taken from post  https://github.com/idaholab/moose/discussions/22326
`mamba activate base    # get back to the original environment`
`mamba create -n petsc moose-petsc moose-tools`
`mamba activate petsc`  
`cd ~/projects/moose`
`git clean -xfd`
`git submodule foreach --recursive git clean -xfd`

`export MOOSE_JOBS=8   # replace this number with how many cores your machine has`
`export METHODS=opt`
`scripts/update_and_rebuild_libmesh.sh --enable-all-static --enable-static`
```

**Reply by milljm** [reply]
If you're going to use Mamba, you might as well create and distribute a mamba package containing your binary instead (since if using Mamba, you'll need to have each of your users also install Mamba).

Have a look at `moose/conda/generate_recipe.sh`. The gist of this tool's use is as follows:

```bash
cd ~/projects
git clone https://github.com/idaholab/moose
git clone <your moose-based application>
export MOOSE_SKIP_DOCS=True
moose/conda/generate_recipe.sh /absolute/path/to/<your moose-based application> 12
```
- _12 being the number of cores you can use on this machine_
- _MOOSE_SKIP_DOCS=True_ telling the `make install` process to ignore building documentation if you prefer

This will pre-install everything necessary (including Conda, boa, etc) into a temporary directory, and proceed to building your application as a relocatable package (A conda package). Much like; `moose-libmesh, moose-tools, etc`.

Once/If successful, your package can be installed by others, via:

```bash
mamba install /path/to/created/conda-package.tar.gz
```
